fuel-block-producer/src/block_producer.rs

```diff
@@ -0,0 +1,194 @@
+    debug,
+    error,
+};
+
+#[cfg(test)]
+mod tests;
+mod transaction_selector;
+
+pub struct Producer<'a> {
+    pub config: Config,
+    pub db: &'a dyn BlockProducerDatabase,
+    pub txpool: &'a dyn TxPool,
+    pub executor: &'a dyn Executor,
+    pub relayer: &'a dyn Relayer,
+}
+
+#[async_trait::async_trait]
+impl<'a> Trait for Producer<'a> {
+    /// Produces a block for the specified height
+    async fn produce_block(&self, height: BlockHeight) -> Result<FuelBlock> {
+        //  - get previous block info (hash, root, etc)
+        //  - select best da_height from relayer
+        //  - get available txs from txpool
+        //  - select best txs based on factors like:
+        //      1. fees
+        //      2. parallel throughput
+        //  - Execute block with production mode to correctly malleate txs outputs and block headers
+
+        let previous_block_info = self.previous_block_info(height).await?;
+        let new_da_height = self
+            .select_new_da_height(previous_block_info.da_height)
+            .await?;
+
+        // transaction selection could use a plugin based approach in the
+        // future for block producers to customize block building (e.g. alternative priorities besides gas fees)
+        let best_transactions = self.select_best_transactions(height).await?;
+
+        let header = FuelBlockHeader {
+            height,
+            number: new_da_height,
+            parent_hash: previous_block_info.hash,
+            // TODO: this needs to be updated using a proper BMT MMR
+            prev_root: previous_block_info.prev_root,
+            // This will be set by the executor
+            transactions_root: Default::default(),
+            time: Utc::now(),
+            // TODO: is this field required?
+            producer: self.config.validator_id,
+            metadata: None,
+        };
+        let mut block = FuelBlock {
+            header,
+            transactions: best_transactions.into_iter().map(Into::into).collect(),
+        };
+        let result = self
+            .executor
+            .execute(&mut block, ExecutionMode::Production)
+            .await;
+
+        if let Err(
+            Error::VmExecution { transaction_id, .. }
+            | Error::TransactionIdCollision(transaction_id),
+        ) = &result
+        {
+            // TODO: if block execution fails due to any transaction validity errors,
+            //          should those txs be removed from the txpool? While this
+            //          theoretically shouldn't happen due to txpool validation rules,
+            //          it is a possibility.
+            error!(
+                "faulty tx prevented block production: {:#x}",
+                transaction_id
+            );
+        }
+
+        let _ = result.context(format!(
+            "Failed to produce block {:?} due to execution failure",
+            block
+        ))?;
+
+        debug!("Produced block: {:?}", &block);
+        Ok(block)
```

Comment:
When do we plan to sign it? Seems BlockProducer should have `Signer` and sign it

Reply:
fixed by removing producer identity from the config. This will be deferred to the consensus module.